Add Jumplist 'Run as Administrator' to Taskbar on Windows

bergmeister opened the pull request:
## PR Summary

~~**travisez13: marking as WIP while I run compliance tools to make sure this doesn't break any compliance rules.  Don't remove WIP until I've signed off on compliance issues**~~

Closes #6649 

This is a port of existing C++ Windows PowerShell code from [MainEntry.cpp](https://github.com/PowerShell/PowerShell/blob/master/src/powershell-native/nativemsh/pwrshexe/MainEntry.cpp)
Some of the code has been copied and minified from the [WindowsApiPack](https://github.com/contre/Windows-API-Code-Pack-1.1).

The code is not compiled for `Linux` (not sure also another condition is needed for Windows on ARM?).
The code checks if the PowerShell process has a window handle by checking the startupinfo and only then tries to populate the list (and also checks if there is a slot available in the jumplist).
Tested on Windows 10 1803, jumpLists have been supported in Windows since Windows 7, which matches what PowerShell Core supports.

This is how it looks like:

![image](https://user-images.githubusercontent.com/9250262/39961300-e2e43a0c-562a-11e8-972f-1821ad1cf621.png)

Note that I found that on WinServer 2012 R2, the icon does not recognised and is empty instead, I tried using the `SetIconLocation` method to point it to the ico file instead (which works on Win10) but it did not work on WinServer2012R2, therefore I suspect I have either an odd VM or it is an OS bug. Since this is only a minor optical and non-functional detail on a less common OS, I think this is OK.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [x] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

bergmeister committed:
First working version of minimalistic elevated Jumplist

bergmeister committed:
add copyright header

bergmeister committed:
Fix CodeFactor warnings and cleanup

bergmeister committed:
remove redundant, leftover TODO item that has been done

bergmeister committed:
make as much internal as possible

bergmeister committed:
Use resource string for 'Run as Administrator'

bergmeister committed:
remove last TODO

bergmeister committed:
[feature] Use proper conditional compilation for Windows only as per PR review.

bergmeister commented on src/Microsoft.PowerShell.ConsoleHost/Microsoft.PowerShell.ConsoleHost.csproj:
```diff
@@ -16,6 +16,14 @@
     <DefineConstants>$(DefineConstants);CORECLR</DefineConstants>
   </PropertyGroup>
 
+  <ItemGroup Condition=" '$(Configuration)' == 'Linux' ">
```
Thanks, I addressed that.

bergmeister commented:
@TravisEz13 Thanks for having a quick look. I finished the last cleanup items and it is not WIP any more.

bergmeister commented:
@TravisEz13 By 'registered', do you mean if the jumplist has already been populated? As far as I am aware the Windows APIs for the jumplist, do not have 'Get' methods to check their current state (but I might be wrong). Any pointers are welcome.
There is no hurry for the review although getting it into the next preview might be good for testing exposure. So take your time, I will be on a conference from Wednesday on anyway.

iSazonov commented on src/Microsoft.PowerShell.ConsoleHost/WindowsTaskbarJumpList/TaskbarJumpList.cs:
```diff
@@ -0,0 +1,69 @@
+        {
+            if (hResult < 0)
+            {
+                throw new Exception($"HResult from COM call was negative, which indicates a failure. Was: '{hResult}'");
```
We use explicit formating in our code and never use the operator `$()`.

Is the message is user-faced? If so we should move it in resource file.

bergmeister replied:
Good point. I was thinking about this more after I opened the PR. Initially I wrote it as a check to make sure all COM calls were successful at each stage during development. Looking at it again now, it is probably better to not throw but instead make the function return/stop prematurely (which should not be a problem because we haven't commited the transaction)

iSazonov commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs:
```diff
@@ -1350,7 +1353,7 @@ internal TextWriter ConsoleTextWriter
         private uint Run(CommandLineParameterParser cpp, bool isPrestartWarned)
         {
-            Dbg.Assert(null != cpp, "CommandLine parameter parser cannot be null.");
+            Dbg.Assert(cpp != null, "CommandLine parameter parser cannot be null.");
```
Should we replace the assert with throw?

bergmeister replied:
No strong opinion on this one, this line was only changed because the code style checker flagged it up that null was on the left hand side (which is only necessary in C/C++ but not C#)

bergmeister committed:
improve error handling: abort jumplist creation and use debug.assert instead of throwing an exception. Since this now not a user facing string any more, I do not think it is worth having the StringUtil.Format wrapper for it

iSazonov commented:
Does that slow down the PowerShell Core start?

bergmeister commented:
I have not measured it but I could not feel a difference. It also only affects pwsh when it is being opened in an actual window.

iSazonov commented:
I guess we get this easily with .Net Core 3.0 - right?

bergmeister commented:
Maybe, but .Net Core 3 development has only just started and realistically speaking, the chance that they will  be delayed is high and they are probably going to deliver a MVP first. But it is not a reason against this PR because one can just delete the folder that I added, so there is no deep integration that would make it difficult to take out in the future

iSazonov commented:
.Net Framework has [System.Windows.Shell.JumpList ](https://msdn.microsoft.com/en-us/library/system.windows.shell.jumplist(v=vs.110).aspx)

bergmeister commented:
Yes, but it does not support elevation.

bergmeister commented:
I could give you a measurement using the Stopwatch class when being compiled in Release and CrossGen'd to give you a rough number. Hardware would be either a 7th Gen i7 Notebook processor or an Azure VM size of your choice.
If you are worried, then we could potentially also implement the option of disabling the feature if a special environment variable is set. Otherwise it is a case of 'there is no such thing as free lunch'.

iSazonov commented:
The code remove one extra mouse click in GUI. This seems to be a very expensive solution, given that there are many non-interactive scenarios.

bergmeister commented:
Expensive in terms of what? In non-interactive scenarios, the only overhead is one call to get the startupinfo and then the algorithm will not proceed.

iSazonov commented:
> Expensive in terms of what?

Over 600 code lines to add one menu item.

bergmeister commented:
The main code is just the 90 lines in the TaskbarJumpList file, which was a port of the existing Windows PowerShell C++ code in the referenced MainEntry.cpp
The other classes were taken from the WindowsApiPack and already minimized to include only the smallest set of required classes and methods
Is the concern more about the review or maintenance/support of this PR?

iSazonov commented:
@bergmeister I'm trying to understand the need for this change, the difficulty in supporting and the negative impact on the start. So far I can take it but I don't see much benefits.

bergmeister commented:
The benefit of this feature is that it halves the time to open am elevated shell, brings consistency with Windows PowerShell and having one less click to do reduces muscle strain in the long term.

The difficulty in writing/supporting this feature is the lack of support for this scenario from the Microsoft/.Net side:
The COM APIs are not very user friendly and require memory structs that are awkard to work with in .Net (especially `PropVariant`). Therefore the Windows team made the WindowsApiPack in .Net 3.5 and stopped supporting it because it got merged into WPF, but both implementations are missing overloads to support elevation and do not provide access to the low level shell object pointer)
Hence why I ripped out the minimum necessary code for us and added a call to the COM Api that enables elevation. But the whole algorithm is the same that Windows PowerShell uses as well.

bergmeister committed:
Merge branch 'master' of https://github.com/PowerShell/PowerShell into JumpListStartAsAdmin_COM

bergmeister commented:
@TravisEz13 @iSazonov @SteveL-MSFT I measured the performance impact using the `StopWatch` class: The first time it took 70 ms and after that the times were around **55 ms**. I did notice however that the times were the same when being launched in a new window or in the same windows. By adding debug statements I found that this is due to the startupInfo check not being sufficient to make it stop early. I think the last case could be improved to reduce the impact for starting new non-interactive runs on Windows or when launching pwsh as part of an existing shell window. 

- OS: Windows 10 1803 (10.0.17134 Build 17134)
- CPU: [Intel i7-7700HQ](https://ark.intel.com/products/97185/Intel-Core-i7-7700HQ-Processor-6M-Cache-up-to-3_80-GHz) (4 core Laptop version with 45 W TDP)
- Disk: SSD - Samsung EVO 850

bergmeister committed:
Fix StartupInfo check to skip calls when shell is not owning the window (and exit even earlier when being in a non-interactive thread)
This reduces the overhead to less than 1ms when the shell is not owning the current window

bergmeister commented on src/Microsoft.PowerShell.ConsoleHost/WindowsTaskbarJumpList/TaskbarJumpList.cs:
```diff
@@ -12,83 +12,87 @@ internal static class TaskbarJumpList
-            GetStartupInfo(out StartUpInfo startupInfo);
-            var STARTF_TITLEISLINKNAME = 0x00000800;
-            if (startupInfo.lpTitle == null || (startupInfo.dwFlags & STARTF_TITLEISLINKNAME) != STARTF_TITLEISLINKNAME)
+            if (Environment.UserInteractive)
```
Lol. I will remove it then. The idea behind this was to not even have a single COM call in non-interactive mode (but it will should stop after the call to getstartupinfo then)

iSazonov replied:
I think it should be PowerShell internal flag - whether it interacts with an user.

bergmeister committed:
Remove redundant call to Environment.UserInteractive (that always returns true). In non-interactive mode the algorithm will stop after the call to getstartupinfo then

bergmeister committed:
Simplify conditional compile option in csproj for Windows

iSazonov commented:
@bergmeister Thanks for your contribution! I am delighted that you have found this solution.
